### DISCUSSION TITLE
Converting Ontonotes ( .gold_conll file format) to DocBins ( .spacy file format)

### DISCUSSION THREAD

**Discussion by emiltj** [explosion/spaCy#12222] [Help: Coding & Implementations] [answered]
Hi!

TLDR; How to convert .gold_conll (as seen in table below) to .spacy?

```
#begin document (bc/cctv/00/cctv_0001); part 000
bc/cctv/00/cctv_0001   0    6               a    DT      (NP(NP*        -    -   -   Speaker#1       *             *    (ARG1*   -
bc/cctv/00/cctv_0001   0    7         special    JJ            *        -    -   -   Speaker#1       *             *         *   -
bc/cctv/00/cctv_0001   0    8         edition    NN            *)       -    -   -   Speaker#1       *             *         *   -
bc/cctv/00/cctv_0001   0    9              of    IN         (PP*        -    -   -   Speaker#1       *             *         *   -
bc/cctv/00/cctv_0001   0   10          Across   NNP         (NP*        -    -   -   Speaker#1   (ORG*             *         *   -
bc/cctv/00/cctv_0001   0   11           China   NNP      *)))))))       -    -   -   Speaker#1       *)            *)        *)  -

#end document
```

My plan is to work with the OntoNotes data in spaCy. I have retrieved the OntoNotes dataset, however, I am struggling to convert my data to a format. I see that the format is similar, but not identical, to two of the sample data formats that may be converted using the python -m spacy converter CLI-command, both trying the --converter auto, conllu, and conll ([ner-token-per-line.iob](https://github.com/explosion/spaCy/blob/master/extra/example_data/ner_example_data/ner-token-per-line.iob) and [ner-token-per-line-conll2003.iob](https://github.com/explosion/spaCy/blob/master/extra/example_data/ner_example_data/ner-token-per-line-conll2003.iob)). Above is a table of what my data looks like, and [this](https://data.mendeley.com/datasets/zmycy7t9h9/2) is where I have retrieved it from. I have been unable to find the entire dataset anywhere else, and I see that none of the [HuggingFace datasets](https://huggingface.co/datasets/tner/ontonotes5) are available with 1) all the data, and 2) in the right format.

**Comment by rmitsch** [comment, marked as answer]
Hi @emiltj, have you tried using the `-c` option to specify the `conll` converter? See the docs [here](https://spacy.io/api/cli#converters).
E. g.: `spacy convert nameofyourfile.conll -c conll ./output/`

**Reply by emiltj** [reply]
Hi rmitsch,

Thank you for your answer. I have tried that, a long with the --converter conllu as well as all other --converter specifications. Any other ideas?

Best

**Reply by rmitsch** [reply]
What's the output you're getting? Can you share a sample of the dataset you're using?

**Reply by emiltj** [reply]
Yes, of course. Also thank you very much for helping out!

**The dataset:**
[Extended version of OntoNotesV5.0 used in CoNLL2012](https://data.mendeley.com/datasets/zmycy7t9h9/2) (the conll-2012.zip file)

**The file:**
Can be downloaded here (I uploaded for convenience) https://we.tl/t-lUMJKexB4A
Or can be found in this folder (after unzipping the conll-2012.zip file from the dataset link above): conll-2012/v12/data/train/data/english/annotations/bc/cctv/00 

**The dataset looks like this:**
```
#begin document (bc/cctv/00/cctv_0001); part 000
bc/cctv/00/cctv_0001   0   0      What   WP  (TOP(SBARQ(WHNP(WHNP*        -   -   -   Speaker#1   *   -
bc/cctv/00/cctv_0001   0   1      kind   NN                      *)       -   -   -   Speaker#1   *   -
bc/cctv/00/cctv_0001   0   2        of   IN                   (PP*        -   -   -   Speaker#1   *   -
bc/cctv/00/cctv_0001   0   3    memory   NN                 (NP*)))   memory  -   1   Speaker#1   *   -
bc/cctv/00/cctv_0001   0   4         ?    .                     *))       -   -   -   Speaker#1   *   -

bc/cctv/00/cctv_0001   8   0          This    DT  (TOP(S(NP*           -    -   -   Speaker#6   *      *        (ARG1*   (37
bc/cctv/00/cctv_0001   8   1          case    NN           *)          -    -   -   Speaker#6   *      *             *)   37)
bc/cctv/00/cctv_0001   8   2          will    MD        (VP*           -    -   -   Speaker#6   *      *    (ARGM-MOD*)    -
bc/cctv/00/cctv_0001   8   3           not    RB           *           -    -   -   Speaker#6   *      *    (ARGM-NEG*)    -
bc/cctv/00/cctv_0001   8   4            be    VB        (VP*           be  03   -   Speaker#6   *    (V*)            *     -
bc/cctv/00/cctv_0001   8   5    prosecuted   VBN      (VP*)))   prosecute  01   1   Speaker#6   *      *           (V*)    -
bc/cctv/00/cctv_0001   8   6             .     .          *))          -    -   -   Speaker#6   *      *             *     -

#end document

```

**When using the `python -m spacy convert data/ontonotes/v12/cctv_0005.gold_conll . --converter conll`, I get:**
```
ℹ Grouping every 1 sentences into a document.
⚠ To generate better training data, you may want to group sentences
into documents with `-n 10`.
Traceback (most recent call last):
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/cli/_util.py", line 74, in setup_cli
    command(prog_name=COMMAND)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/typer/main.py", line 532, in wrapper
    return callback(**use_params)  # type: ignore
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/cli/convert.py", line 77, in convert_cli
    convert(
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/cli/convert.py", line 142, in convert
    db = DocBin(docs=docs, store_user_data=True)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/tokens/_serialize.py", line 87, in __init__
    for doc in docs:
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/training/converters/conll_ner_to_docs.py", line 109, in conll_ner_to_docs
    biluo_tags.extend(iob_to_biluo(cols[-1]))
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/training/iob_utils.py", line 13, in iob_to_biluo
    out.extend(_consume_ent(tags))
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/training/iob_utils.py", line 46, in _consume_ent
    raise ValueError(Errors.E177.format(tag=tag))
ValueError: [E177] Ill-formed IOB input detected: :
```

**When using the `python -m spacy convert data/ontonotes/v12/cctv_0005.gold_conll . --converter conllu`, I get:**
```
ℹ Grouping every 1 sentences into a document.
⚠ To generate better training data, you may want to group sentences
into documents with `-n 10`.
Traceback (most recent call last):
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/cli/_util.py", line 74, in setup_cli
    command(prog_name=COMMAND)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/typer/main.py", line 532, in wrapper
    return callback(**use_params)  # type: ignore
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/cli/convert.py", line 77, in convert_cli
    convert(
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/cli/convert.py", line 142, in convert
    db = DocBin(docs=docs, store_user_data=True)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/tokens/_serialize.py", line 87, in __init__
    for doc in docs:
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/training/converters/conllu_to_docs.py", line 38, in conllu_to_docs
    for sent_doc in sent_docs:
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/training/converters/conllu_to_docs.py", line 74, in read_conllx
    set_ents = has_ner(input_data, ner_tag_pattern)
  File "/Users/emiltrencknerjessen/miniconda3/envs/thesis/lib/python3.10/site-packages/spacy/training/converters/conllu_to_docs.py", line 58, in has_ner
    id_, word, lemma, pos, tag, morph, head, dep, _1, misc = parts
ValueError: not enough values to unpack (expected 10, got 1)
```

**Reply by rmitsch** [reply]
Thanks for the information. Backing up a bit: what are you attempting to do with this data?

After looking into this, the `conll` converter supports the the _conll-2003_ format, which is different from the one you're using. This means you'll have to convert this into Doc objects (or some other supported format) yourself.

**Reply by emiltj** [reply]
> Backing up a bit: what are you attempting to do with this data?

I want to load the Ontonotes dataset as spacy Doc objects, so I can work with it in spaCy. 
I only want to use the NE tags and the text itself.

> After looking into this, the conll converter supports the the conll-2003 format, which is different from the one you're using. This means you'll have to convert this into Doc objects (or some other supported format) yourself.

I also suspected that the format I am using is not the format that the convert expects. However, I don't know how to manually convert this file to a .spacy file.

I see that many of spaCys models (e.g. [en_core_web_sm](https://spacy.io/models/en#en_core_web_sm) are trained using the Ontonotes dataset. Before this, I have therefore gained access to the dataset using the [link provided in spaCy documentation, e.g. under the en_core_model_sm
](https://catalog.ldc.upenn.edu/LDC2013T19). However, these files do not look like the conll-2003 format that is documented in the [converters documentation](https://spacy.io/api/cli#converters), either.

This might not be the correct forum for this, but is there any way of getting the Ontonotes dataset in any sort of spaCy format?

**Reply by adrianeboyd** [reply]
This is the official release of OntoNotes that is used for the trained pipelines: https://catalog.ldc.upenn.edu/LDC2013T19

Unfortunately our OntoNotes license does not permit us to redistribute any of the full texts or annotations. I don't know the details for sure, but I don't think that the version in that link is allowed to redistribute the texts or release it under a CC BY 4.0 license. The annotation columns were made available for the CoNLL 2012 shared task (https://conll.cemantix.org/2012/data.html), but not the texts.

With the same idea as in that shared task, we have considered developing scripts for users to re-create our internal `en_core_web_*` training corpus from the original OntoNotes release, but it would be a big effort on our side without a huge benefit, since most users don't have access to OntoNotes.

In any case, just looking at the data format, you want to convert the text column (column 4) and the NE column (column 11) from this to something spacy can use directly or convert. I'd recommend converting to CoNLL-2003 format, which is what `spacy convert -c conll` supports. For the spacy converter you can skip the middle two columns and just provide the first and last columns.

This annotation:

```
a       *
b       (PERSON*
c       *)
d       *
e       *
```

would become:

```
a       O
b       B-PERSON
c       I-PERSON
d       O
e       O
```

**Reply by emiltj** [reply]
Hi @adrianeboyd ,

Thank you very much for your thorough answer. I have also gained access using the legitimate channel, but since these files seemingly weren't in the right format either, I have tried this more untraditional way of accessing the data, since it looked more akin to the conll format.

But just to clarify:
The documentation states that the conll-2003 format should be formatted like this:
> <img width="696" alt="image" src="https://user-images.githubusercontent.com/54767532/216806850-7a982cc8-ffc8-4ac4-ba04-6c574d062b89.png">
Which matches[ the examples given](https://github.com/explosion/spaCy/blob/master/extra/example_data/ner_example_data/ner-token-per-line-conll2003.iob) in the documentation that looks like this:
```
-DOCSTART- -X- O O

When WRB _ O
Sebastian NNP _ B-PERSON
Thrun NNP _ I-PERSON
started VBD _ O
working VBG _ O
```

However, the format you suggest matches [another example](https://github.com/explosion/spaCy/blob/master/extra/example_data/ner_example_data/ner-token-per-line.iob), but does not match the description of conll-2003 from [the documentation](https://spacy.io/api/cli#converters).
```
a       O
b       B-PERSON
c       I-PERSON
d       O
e       O
```

- Q1: Is the latter format (right above this text), also a case of conll-2003? And may it be converted using the convert -c conll, despite it not matching [the documentation](https://spacy.io/api/cli#converters)?
- Q2: You mentioned that "[...] _You want to convert the text column (column 4) and the NE column (column 11) from this to something spacy can use directly or convert_".  and "_You can skip the middle two columns and just provide the first and last columns._". Does this mean I can filter to only keep col 4 (as is), and col 11 (converted), and then the `convert` will work?

Best

**Reply by adrianeboyd** [reply]
The format quoted above is not exactly CoNLL-2003, but the converter is written to be a little more flexible for similar formats and includes CoNLL-2003 in what it supports. It works as described in the docs. Note that the doc start line in the spacy converter is only that exact string (spaces not tabs, no alternate `-X-`/`O` values) and both the doc and sentence breaks are optional. If your data doesn't have doc or sentence breaks, you could use `-s` or `-n` to automatically add them.

Please just try it out and double-check by loading your `.spacy` files with `DocBin` to check that the annotation and sentence/document boundaries are where you'd expect so you can catch any problems before training.

**Reply by emiltj** [reply]
Okay,

Thank you very much for the help. 

Good day to you!
